File: packages/core/src/db/cards.ts
import type { Database } from "bun:sqlite";
import type {
  CardId,
  CardWithTags,
  CreateCard,
  UpdateCard,
  MoveCard,
  ReorderCards,
  BoardId,
  CardStatusType,
} from "../types/index.js";
import { CreateCardSchema } from "../schemas/card.js";

interface CardRow {
  id: string;
  board_id: string;
  title: string;
  description: string;
  status: string;
  position: number;
  blocking: number;
  plan_thinking: string | null;
  execute_thinking: string | null;
  auto_commit: number | null;
  auto_push: number | null;
  cli_provider: string | null;
  cli_custom_command: string | null;
  branch_mode: string | null;
  branch_name: string | null;
  assignee: string;
  created_at: string;
  updated_at: string;
}

function getTagsForCard(db: Database, cardId: string): string[] {
  const rows = db
    .query("SELECT tag FROM card_tags WHERE card_id = ? ORDER BY tag")
    .all(cardId) as Array<{ tag: string }>;
  return rows.map((r) => r.tag);
}

function setTagsForCard(
  db: Database,
  cardId: string,
  tags: string[]
): void {
  db.query("DELETE FROM card_tags WHERE card_id = ?").run(cardId);
  const stmt = db.query(
    "INSERT INTO card_tags (card_id, tag) VALUES (?, ?)"
  );
  for (const tag of tags) {
    stmt.run(cardId, tag);
  }
}

function getFilesForCard(db: Database, cardId: string): string[] {
  const rows = db
    .query("SELECT file_path FROM card_files WHERE card_id = ? ORDER BY file_path")
    .all(cardId) as Array<{ file_path: string }>;
  return rows.map((r) => r.file_path);
}

function setFilesForCard(
  db: Database,
  cardId: string,
  files: string[]
): void {
  db.query("DELETE FROM card_files WHERE card_id = ?").run(cardId);
  const stmt = db.query(
    "INSERT INTO card_files (card_id, file_path) VALUES (?, ?)"
  );
  for (const filePath of files) {
    stmt.run(cardId, filePath);
  }
}

function toCardWithTags(db: Database, row: CardRow): CardWithTags {
  return {
    ...row,
    blocking: Boolean(row.blocking),
    plan_thinking: row.plan_thinking as "smart" | "basic" | null,
    execute_thinking: row.execute_thinking as "smart" | "basic" | null,
    auto_commit: row.auto_commit === null ? null : row.auto_commit !== 0,
    auto_push: row.auto_push === null ? null : row.auto_push !== 0,
    cli_provider: (row.cli_provider ?? null) as CardWithTags["cli_provider"],
    cli_custom_command: row.cli_custom_command ?? null,
    branch_mode: (row.branch_mode ?? null) as CardWithTags["branch_mode"],
    branch_name: row.branch_name ?? null,
    assignee: (row.assignee ?? "ai") as "ai" | "human",
    tags: getTagsForCard(db, row.id),
    files: getFilesForCard(db, row.id),
  } as CardWithTags;
}

export function listCards(
  db: Database,
  boardId: BoardId,
  options?: { doneLimit?: number }
): { cards: CardWithTags[]; doneHasMore: boolean } {
  const doneLimit = options?.doneLimit ?? 20;

  const nonDoneRows = db
    .query(
      "SELECT * FROM cards WHERE board_id = ? AND status != 'done' ORDER BY position ASC, created_at ASC"
    )
    .all(boardId) as CardRow[];

  const doneRows = db
    .query(
      "SELECT * FROM cards WHERE board_id = ? AND status = 'done' ORDER BY updated_at DESC LIMIT ?"
    )
    .all(boardId, doneLimit + 1) as CardRow[];

  const doneHasMore = doneRows.length > doneLimit;
  const limitedDoneRows = doneRows.slice(0, doneLimit);

  const cards = [...nonDoneRows, ...limitedDoneRows].map((row) =>
    toCardWithTags(db, row)
  );
  return { cards, doneHasMore };
}

export function listCardsByStatus(
  db: Database,
  boardId: BoardId,
  status: CardStatusType
): CardWithTags[] {
  const rows = db
    .query(
      "SELECT * FROM cards WHERE board_id = ? AND status = ? ORDER BY position ASC, created_at ASC"
    )
    .all(boardId, status) as CardRow[];
  return rows.map((row) => toCardWithTags(db, row));
}

export function getCard(db: Database, id: CardId): CardWithTags | null {
  const row = db
    .query("SELECT * FROM cards WHERE id = ?")
    .get(id) as CardRow | null;
  if (!row) return null;
  return toCardWithTags(db, row);
}

export function createCard(
  db: Database,
  boardId: BoardId,
  rawInput: CreateCard
): CardWithTags {
  const input = CreateCardSchema.parse(rawInput);
  // Get next position if not specified
  const status = input.status ?? "todo";
  const position =
    input.position ??
    ((
      db
        .query(
          "SELECT COALESCE(MAX(position), -1) + 1 as next_pos FROM cards WHERE board_id = ? AND status = ?"
        )
        .get(boardId, status) as { next_pos: number }
    ).next_pos);

  const row = db
    .query(
      `INSERT INTO cards (board_id, title, description, status, position, blocking, plan_thinking, execute_thinking, auto_commit, auto_push, cli_provider, cli_custom_command, branch_mode, branch_name, assignee)
       VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)
       RETURNING *`
    )
    .get(boardId, input.title, input.description, status, position, input.blocking ? 1 : 0, input.plan_thinking ?? null, input.execute_thinking ?? null, input.auto_commit === undefined ? null : input.auto_commit === null ? null : (input.auto_commit ? 1 : 0), input.auto_push === undefined ? null : input.auto_push === null ? null : (input.auto_push ? 1 : 0), input.cli_provider ?? null, input.cli_custom_command ?? null, input.branch_mode ?? null, input.branch_name ?? null, input.assignee ?? "ai") as CardRow;

  if (input.tags.length > 0) {
    setTagsForCard(db, row.id, input.tags);
  }
  if (input.files.length > 0) {
    setFilesForCard(db, row.id, input.files);
  }

  return toCardWithTags(db, row);
}

export function updateCard(
  db: Database,
  id: CardId,
  input: UpdateCard
): CardWithTags | null {
  const current = db
    .query("SELECT * FROM cards WHERE id = ?")
    .get(id) as CardRow | null;
  if (!current) return null;

  const title = input.title ?? current.title;
  const description = input.description ?? current.description;
  const status = input.status ?? current.status;
  const position = input.position ?? current.position;
  const blocking = input.blocking !== undefined ? (input.blocking ? 1 : 0) : current.blocking;
  const planThinking = input.plan_thinking !== undefined ? input.plan_thinking : current.plan_thinking;
  const executeThinking = input.execute_thinking !== undefined ? input.execute_thinking : current.execute_thinking;
  const autoCommit = input.auto_commit !== undefined
    ? (input.auto_commit === null ? null : (input.auto_commit ? 1 : 0))
    : current.auto_commit;
  const autoPush = input.auto_push !== undefined
    ? (input.auto_push === null ? null : (input.auto_push ? 1 : 0))
    : current.auto_push;
  const cliProvider = input.cli_provider !== undefined ? input.cli_provider : current.cli_provider;
  const cliCustomCommand = input.cli_custom_command !== undefined ? input.cli_custom_command : current.cli_custom_command;
  const branchMode = input.branch_mode !== undefined ? input.branch_mode : current.branch_mode;
  const branchName = input.branch_name !== undefined ? input.branch_name : current.branch_name;
  const assignee = input.assignee ?? current.assignee;

  const row = db
    .query(
      `UPDATE cards SET title = ?, description = ?, status = ?, position = ?, blocking = ?, plan_thinking = ?, execute_thinking = ?, auto_commit = ?, auto_push = ?, cli_provider = ?, cli_custom_command = ?, branch_mode = ?, branch_name = ?, assignee = ?, updated_at = datetime('now')
       WHERE id = ?
       RETURNING *`
    )
    .get(title, description, status, position, blocking, planThinking, executeThinking, autoCommit, autoPush, cliProvider, cliCustomCommand, branchMode, branchName, assignee, id) as CardRow;

  if (input.tags !== undefined) {
    setTagsForCard(db, row.id, input.tags);
  }
  if (input.files !== undefined) {
    setFilesForCard(db, row.id, input.files);
  }

  return toCardWithTags(db, row);
}

export function moveCard(
  db: Database,
  id: CardId,
  move: MoveCard
): CardWithTags | null {
  const row = db
    .query(
      `UPDATE cards SET status = ?, position = ?, updated_at = datetime('now')
       WHERE id = ?
       RETURNING *`
    )
    .get(move.status, move.position, id) as CardRow | null;
  if (!row) return null;
  return toCardWithTags(db, row);
}

export function reorderCards(db: Database, updates: ReorderCards): void {
  const stmt = db.prepare(
    "UPDATE cards SET status = ?, position = ?, updated_at = datetime('now') WHERE id = ?"
  );
  const tx = db.transaction(() => {
    for (const u of updates) {
      stmt.run(u.status, u.position, u.id);
    }
  });
  tx();
}

export function moveCardToBoard(
  db: Database,
  cardId: string,
  targetBoardId: string
): CardWithTags | null {
  const nextPos = (
    db
      .query(
        "SELECT COALESCE(MAX(position), -1) + 1 as next_pos FROM cards WHERE board_id = ? AND status = 'todo'"
      )
      .get(targetBoardId) as { next_pos: number }
  ).next_pos;

  const row = db
    .query(
      `UPDATE cards SET board_id = ?, status = 'todo', position = ?, updated_at = datetime('now')
       WHERE id = ?
       RETURNING *`
    )
    .get(targetBoardId, nextPos, cardId) as CardRow | null;
  if (!row) return null;
  return toCardWithTags(db, row);
}

export function deleteCard(db: Database, id: CardId): boolean {
  const result = db.query("DELETE FROM cards WHERE id = ?").run(id);
  return result.changes > 0;
}

export function updateCardStatus(
  db: Database,
  id: CardId,
  status: CardStatusType
): void {
  db.query(
    "UPDATE cards SET status = ?, updated_at = datetime('now') WHERE id = ?"
  ).run(status, id);
}

export function resetStaleCards(db: Database): number {
  const result = db
    .query(
      "UPDATE cards SET status = 'todo' WHERE status IN ('in-progress', 'queued')"
    )
    .run();
  return result.changes;
}

export function recoverInterruptedCards(db: Database): { requeued: number; reset: number } {
  const inProgressIds = db
    .query("SELECT id FROM cards WHERE status = 'in-progress'")
    .all() as Array<{ id: string }>;

  let requeued = 0;
  let reset = 0;

  for (const { id } of inProgressIds) {
    // Cancel any running executions for this card
    db.query(
      "UPDATE executions SET status = 'cancelled', finished_at = datetime('now') WHERE card_id = ? AND status = 'running'"
    ).run(id);

    // Check if a plan phase completed successfully
    const planDone = db
      .query(
        "SELECT id FROM executions WHERE card_id = ? AND phase = 'plan' AND status = 'success' ORDER BY started_at DESC LIMIT 1"
      )
      .get(id);

    if (planDone) {
      db.query("UPDATE cards SET status = 'queued', updated_at = datetime('now') WHERE id = ?").run(id);
      requeued++;
    } else {
      db.query("UPDATE cards SET status = 'todo', updated_at = datetime('now') WHERE id = ?").run(id);
      reset++;
    }
  }

  return { requeued, reset };
}

export function countActiveCards(db: Database): number {
  const row = db
    .query(
      "SELECT COUNT(*) as count FROM cards WHERE status IN ('queued', 'in-progress')"
    )
    .get() as { count: number };
  return row.count;
}

export function countCardsByStatusAllBoards(
  db: Database
): Record<string, Record<CardStatusType, number>> {
  const rows = db
    .query("SELECT board_id, status, COUNT(*) as count FROM cards GROUP BY board_id, status")
    .all() as Array<{ board_id: string; status: string; count: number }>;

  const result: Record<string, Record<CardStatusType, number>> = {};
  const statuses: CardStatusType[] = ["todo", "queued", "in-progress", "done", "failed"];

  for (const row of rows) {
    if (!result[row.board_id]) {
      result[row.board_id] = { todo: 0, queued: 0, "in-progress": 0, done: 0, failed: 0 };
    }
    if (statuses.includes(row.status as CardStatusType)) {
      result[row.board_id]![row.status as CardStatusType] = row.count;
    }
  }

  return result;
}

export function countDonePerDay(
  db: Database,
  days: number = 14,
  tzOffsetMinutes: number = 0
): Array<{ date: string; count: number }> {
  const modifier = `${-tzOffsetMinutes} minutes`;

  const rows = db
    .query(
      `SELECT date(updated_at, ?) as day, COUNT(*) as count
       FROM cards
       WHERE status = 'done'
         AND date(updated_at, ?) >= date('now', ?, '-' || ? || ' days')
       GROUP BY date(updated_at, ?)
       ORDER BY day ASC`
    )
    .all(modifier, modifier, modifier, days, modifier) as Array<{ day: string; count: number }>;

  const map = new Map(rows.map((r) => [r.day, r.count]));
  const result: Array<{ date: string; count: number }> = [];
  const localNow = new Date(Date.now() - tzOffsetMinutes * 60 * 1000);
  for (let i = days - 1; i >= 0; i--) {
    const d = new Date(localNow);
    d.setUTCDate(d.getUTCDate() - i);
    const key = d.toISOString().slice(0, 10);
    result.push({ date: key, count: map.get(key) ?? 0 });
  }

  return result;
}

export function countDonePerDayByBoard(
  db: Database,
  days: number = 14
): Record<string, Array<{ date: string; count: number }>> {
  const rows = db
    .query(
      `SELECT board_id, date(updated_at) as day, COUNT(*) as count
       FROM cards
       WHERE status = 'done' AND updated_at >= date('now', '-' || ? || ' days')
       GROUP BY board_id, date(updated_at)
       ORDER BY board_id, day ASC`
    )
    .all(days) as Array<{ board_id: string; day: string; count: number }>;

  // Group by board_id
  const byBoard = new Map<string, Map<string, number>>();
  for (const row of rows) {
    if (!byBoard.has(row.board_id)) byBoard.set(row.board_id, new Map());
    byBoard.get(row.board_id)!.set(row.day, row.count);
  }

  // Zero-fill gaps for each board
  const result: Record<string, Array<{ date: string; count: number }>> = {};
  const now = new Date();
  for (const [boardId, map] of byBoard) {
    const series: Array<{ date: string; count: number }> = [];
    for (let i = days - 1; i >= 0; i--) {
      const d = new Date(now);
      d.setDate(d.getDate() - i);
      const key = d.toISOString().slice(0, 10);
      series.push({ date: key, count: map.get(key) ?? 0 });
    }
    result[boardId] = series;
  }

  return result;
}

export function getDistinctTags(db: Database, boardId: BoardId): string[] {
  const rows = db
    .query(
      `SELECT DISTINCT ct.tag FROM card_tags ct
       JOIN cards c ON c.id = ct.card_id
       WHERE c.board_id = ?
       ORDER BY ct.tag`
    )
    .all(boardId) as Array<{ tag: string }>;
  return rows.map((r) => r.tag);
}

File: rust/crates/core/src/db/cards.rs
use rusqlite::{params, Connection, OptionalExtension};

use crate::types::{Card, CardStatus, CardWithTags, CreateCard, UpdateCard};
use crate::Result;

pub fn list_for_board(conn: &Connection, board_id: &str, done_limit: i64) -> Result<(Vec<CardWithTags>, bool)> {
    let mut stmt = conn.prepare(
        "SELECT * FROM cards
         WHERE board_id = ? AND status != 'done'
         ORDER BY position ASC, created_at ASC",
    )?;
    let mut cards: Vec<Card> = stmt
        .query_map([board_id], row_to_card)?
        .collect::<rusqlite::Result<Vec<_>>>()?;

    let mut done_stmt = conn.prepare(
        "SELECT * FROM cards
         WHERE board_id = ? AND status = 'done'
         ORDER BY updated_at DESC
         LIMIT ?",
    )?;
    let done: Vec<Card> = done_stmt
        .query_map(params![board_id, done_limit], row_to_card)?
        .collect::<rusqlite::Result<Vec<_>>>()?;
    cards.extend(done);

    let total_done: i64 = conn.query_row(
        "SELECT COUNT(*) FROM cards WHERE board_id = ? AND status = 'done'",
        [board_id],
        |r| r.get(0),
    )?;
    let done_has_more = total_done > done_limit;

    let with_tags = cards
        .into_iter()
        .map(|c| {
            let tags = tags_for_card(conn, &c.id).unwrap_or_default();
            CardWithTags { card: c, tags }
        })
        .collect();

    Ok((with_tags, done_has_more))
}

pub fn get_with_tags(conn: &Connection, id: &str) -> Result<Option<CardWithTags>> {
    let card = conn
        .query_row("SELECT * FROM cards WHERE id = ?", [id], row_to_card)
        .optional()?;
    match card {
        Some(c) => {
            let tags = tags_for_card(conn, &c.id)?;
            Ok(Some(CardWithTags { card: c, tags }))
        }
        None => Ok(None),
    }
}

pub fn create(conn: &Connection, board_id: &str, input: &CreateCard) -> Result<CardWithTags> {
    let next_position: i64 = conn.query_row(
        "SELECT COALESCE(MAX(position), -1) + 1 FROM cards WHERE board_id = ? AND status = 'todo'",
        [board_id],
        |r| r.get(0),
    )?;

    let assignee = input.assignee.clone().unwrap_or(crate::types::Assignee::Ai);
    let assignee_str = match assignee {
        crate::types::Assignee::Ai => "ai",
        crate::types::Assignee::Human => "human",
    };

    let card = conn.query_row(
        "INSERT INTO cards (board_id, title, description, position, assignee)
         VALUES (?, ?, ?, ?, ?)
         RETURNING *",
        params![
            board_id,
            input.title,
            input.description.clone().unwrap_or_default(),
            next_position,
            assignee_str
        ],
        row_to_card,
    )?;

    if let Some(tags) = &input.tags {
        for tag in tags {
            conn.execute(
                "INSERT OR IGNORE INTO card_tags (card_id, tag) VALUES (?, ?)",
                params![card.id, tag],
            )?;
        }
    }

    let tags = tags_for_card(conn, &card.id)?;
    Ok(CardWithTags { card, tags })
}

pub fn update(conn: &Connection, id: &str, input: &UpdateCard) -> Result<Option<CardWithTags>> {
    let Some(current) = get_with_tags(conn, id)? else {
        return Ok(None);
    };
    let c = &current.card;

    let title = input.title.clone().unwrap_or_else(|| c.title.clone());
    let description = input.description.clone().unwrap_or_else(|| c.description.clone());
    let status = input.status.clone().unwrap_or_else(|| c.status.clone());
    let assignee = input.assignee.clone().unwrap_or_else(|| c.assignee.clone());

    let status_str = status_to_str(&status);
    let assignee_str = match assignee {
        crate::types::Assignee::Ai => "ai",
        crate::types::Assignee::Human => "human",
    };

    conn.execute(
        "UPDATE cards SET title = ?, description = ?, status = ?, assignee = ?, updated_at = datetime('now') WHERE id = ?",
        params![title, description, status_str, assignee_str, id],
    )?;

    if let Some(new_tags) = &input.tags {
        conn.execute("DELETE FROM card_tags WHERE card_id = ?", [id])?;
        for tag in new_tags {
            conn.execute(
                "INSERT OR IGNORE INTO card_tags (card_id, tag) VALUES (?, ?)",
                params![id, tag],
            )?;
        }
    }

    get_with_tags(conn, id)
}

pub fn move_card(conn: &Connection, id: &str, status: CardStatus, position: i64) -> Result<Option<CardWithTags>> {
    let status_str = status_to_str(&status);
    conn.execute(
        "UPDATE cards SET status = ?, position = ?, updated_at = datetime('now') WHERE id = ?",
        params![status_str, position, id],
    )?;
    get_with_tags(conn, id)
}

/// List all cards for a board with a given status, ordered by position then created_at.
pub fn list_by_status(conn: &Connection, board_id: &str, status: CardStatus) -> Result<Vec<CardWithTags>> {
    let status_str = status_to_str(&status);
    let mut stmt = conn.prepare(
        "SELECT * FROM cards WHERE board_id = ? AND status = ? ORDER BY position ASC, created_at ASC",
    )?;
    let cards: Vec<Card> = stmt
        .query_map(params![board_id, status_str], row_to_card)?
        .collect::<rusqlite::Result<Vec<_>>>()?;
    let with_tags = cards
        .into_iter()
        .map(|c| {
            let tags = tags_for_card(conn, &c.id).unwrap_or_default();
            CardWithTags { card: c, tags }
        })
        .collect();
    Ok(with_tags)
}

pub fn set_status(conn: &Connection, id: &str, status: CardStatus) -> Result<()> {
    let status_str = status_to_str(&status);
    conn.execute(
        "UPDATE cards SET status = ?, updated_at = datetime('now') WHERE id = ?",
        params![status_str, id],
    )?;
    Ok(())
}

pub fn count_active(conn: &Connection) -> Result<i64> {
    let n: i64 = conn.query_row(
        "SELECT COUNT(*) FROM cards WHERE status IN ('queued', 'in-progress')",
        [],
        |r| r.get(0),
    )?;
    Ok(n)
}

pub fn delete(conn: &Connection, id: &str) -> Result<bool> {
    let n = conn.execute("DELETE FROM cards WHERE id = ?", [id])?;
    Ok(n > 0)
}

pub fn set_plan_summary(conn: &Connection, id: &str, summary: Option<&str>) -> Result<()> {
    conn.execute(
        "UPDATE cards SET plan_summary = ?, updated_at = datetime('now') WHERE id = ?",
        params![summary, id],
    )?;
    Ok(())
}

pub fn set_completion_summary(conn: &Connection, id: &str, summary: Option<&str>) -> Result<()> {
    conn.execute(
        "UPDATE cards SET completion_summary = ?, updated_at = datetime('now') WHERE id = ?",
        params![summary, id],
    )?;
    Ok(())
}

pub fn set_blocker(conn: &Connection, id: &str, blocker: Option<&str>) -> Result<()> {
    conn.execute(
        "UPDATE cards SET blocker = ?, updated_at = datetime('now') WHERE id = ?",
        params![blocker, id],
    )?;
    Ok(())
}

pub fn clear_blocker(conn: &Connection, id: &str) -> Result<()> {
    set_blocker(conn, id, None)
}

fn tags_for_card(conn: &Connection, card_id: &str) -> Result<Vec<String>> {
    let mut stmt = conn.prepare("SELECT tag FROM card_tags WHERE card_id = ? ORDER BY tag")?;
    let tags: Vec<String> = stmt
        .query_map([card_id], |r| r.get::<_, String>(0))?
        .collect::<rusqlite::Result<Vec<_>>>()?;
    Ok(tags)
}

fn status_to_str(s: &CardStatus) -> &'static str {
    match s {
        CardStatus::Todo => "todo",
        CardStatus::Queued => "queued",
        CardStatus::InProgress => "in-progress",
        CardStatus::Done => "done",
        CardStatus::Failed => "failed",
    }
}

fn row_to_card(row: &rusqlite::Row<'_>) -> rusqlite::Result<Card> {
    let status_str: String = row.get("status")?;
    let status = match status_str.as_str() {
        "todo" => CardStatus::Todo,
        "queued" => CardStatus::Queued,
        "in-progress" => CardStatus::InProgress,
        "done" => CardStatus::Done,
        "failed" => CardStatus::Failed,
        other => return Err(rusqlite::Error::FromSqlConversionFailure(
            0,
            rusqlite::types::Type::Text,
            format!("unknown status: {other}").into(),
        )),
    };
    let assignee_str: String = row.get("assignee")?;
    let assignee = match assignee_str.as_str() {
        "ai" => crate::types::Assignee::Ai,
        "human" => crate::types::Assignee::Human,
        other => return Err(rusqlite::Error::FromSqlConversionFailure(
            0,
            rusqlite::types::Type::Text,
            format!("unknown assignee: {other}").into(),
        )),
    };
    Ok(Card {
        id: row.get("id")?,
        board_id: row.get("board_id")?,
        title: row.get("title")?,
        description: row.get("description")?,
        status,
        position: row.get("position")?,
        blocking: row.get("blocking")?,
        plan_thinking: row.get("plan_thinking")?,
        execute_thinking: row.get("execute_thinking")?,
        auto_commit: row.get("auto_commit")?,
        auto_push: row.get("auto_push")?,
        assignee,
        cli_provider: row.get("cli_provider")?,
        cli_custom_command: row.get("cli_custom_command")?,
        branch_mode: row.get("branch_mode")?,
        branch_name: row.get("branch_name")?,
        plan_summary: row.get("plan_summary")?,
        completion_summary: row.get("completion_summary")?,
        blocker: row.get("blocker")?,
        created_at: row.get("created_at")?,
        updated_at: row.get("updated_at")?,
    })
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::db::{boards, connection::open_memory};
    use crate::types::{CreateBoard, CreateCard};

    fn setup() -> (rusqlite::Connection, String) {
        let conn = open_memory().unwrap();
        let board = boards::create(
            &conn,
            &CreateBoard {
                name: "B".into(),
                description: String::new(),
                directory: "/tmp".into(),
                color: None,
                slug: None,
                github_url: None,
            },
        )
        .unwrap();
        (conn, board.id)
    }

    #[test]
    fn create_assigns_next_position() {
        let (conn, board_id) = setup();
        let a = create(&conn, &board_id, &CreateCard { title: "a".into(), description: None, tags: None, assignee: None }).unwrap();
        let b = create(&conn, &board_id, &CreateCard { title: "b".into(), description: None, tags: None, assignee: None }).unwrap();
        assert_eq!(a.card.position, 0);
        assert_eq!(b.card.position, 1);
    }

    #[test]
    fn tags_round_trip() {
        let (conn, board_id) = setup();
        let card = create(
            &conn,
            &board_id,
            &CreateCard {
                title: "t".into(),
                description: None,
                tags: Some(vec!["x".into(), "y".into()]),
                assignee: None,
            },
        )
        .unwrap();
        assert_eq!(card.tags, vec!["x".to_string(), "y".to_string()]);
    }

    #[test]
    fn update_replaces_tags() {
        let (conn, board_id) = setup();
        let card = create(&conn, &board_id, &CreateCard { title: "t".into(), description: None, tags: Some(vec!["a".into()]), assignee: None }).unwrap();
        let mut patch = UpdateCard::default();
        patch.tags = Some(vec!["b".into(), "c".into()]);
        let updated = update(&conn, &card.card.id, &patch).unwrap().unwrap();
        assert_eq!(updated.tags, vec!["b".to_string(), "c".to_string()]);
    }

    #[test]
    fn move_card_changes_status_and_position() {
        let (conn, board_id) = setup();
        let card = create(&conn, &board_id, &CreateCard { title: "t".into(), description: None, tags: None, assignee: None }).unwrap();
        let moved = move_card(&conn, &card.card.id, CardStatus::Queued, 5).unwrap().unwrap();
        assert_eq!(moved.card.status, CardStatus::Queued);
        assert_eq!(moved.card.position, 5);
    }

    #[test]
    fn count_active_counts_queued_and_in_progress() {
        let (conn, board_id) = setup();
        let a = create(&conn, &board_id, &CreateCard { title: "a".into(), description: None, tags: None, assignee: None }).unwrap();
        let b = create(&conn, &board_id, &CreateCard { title: "b".into(), description: None, tags: None, assignee: None }).unwrap();
        let c = create(&conn, &board_id, &CreateCard { title: "c".into(), description: None, tags: None, assignee: None }).unwrap();
        move_card(&conn, &a.card.id, CardStatus::Queued, 0).unwrap();
        move_card(&conn, &b.card.id, CardStatus::InProgress, 0).unwrap();
        move_card(&conn, &c.card.id, CardStatus::Done, 0).unwrap();
        assert_eq!(count_active(&conn).unwrap(), 2);
    }
}
